CMakeLists.txt

```diff
@@ -34,6 +34,21 @@ if (MSVC)
 	set(objc_LINK_FLAGS "/DEBUG /INCREMENTAL:NO ${objc_LINK_FLAGS}")
 endif()
 
+# Get Architecture without relying on CMake
```

Comment:
Please can you add a comment explaining why we can't rely on CMake here?

Reply:
When using the x64 variant of CMake on Windows (which is the version shipped by Chocolatey), CMake reports that the CPU architecture is AMD64, and not ARM64 as expected.

When using the Preprocessor, we can reliably check the CPU architecture.

Reply:
If you're getting `AMD64` here when cross compiling, I would expect that you haven't set up your cmake configuration entirely correctly for cross compilation; when cross compiling with cmake you need to set `CMAKE_SYSTEM_NAME`, even if cross compiling from Windows x86_64 to Windows aarch64 - see https://cmake.org/cmake/help/latest/variable/CMAKE_CROSSCOMPILING.html and https://cmake.org/cmake/help/latest/variable/CMAKE_SYSTEM_NAME.html.

You didn't say which variable you did check, that gave you `AMD64` - I would expect `CMAKE_SYSTEM_PROCESSOR`. This one is tricky, because when you're not cross compiling, then this is set to a rather inconsistent set of values depending on platform (e.g. amd64/x86_64/x64 and arm64 vs aarch64 and similar, differ depending on OS), see https://cmake.org/cmake/help/latest/variable/CMAKE_HOST_SYSTEM_PROCESSOR.html#variable:CMAKE_HOST_SYSTEM_PROCESSOR. When cross compiling, then CMake doesn't set `CMAKE_SYSTEM_PROCESSOR` at all, but the user configuring the cross compile is supposed to set it. As most things build fine without setting it, many cmake cross compilation setups don't set it at all.

There's another variable, `CMAKE_C_COMPILER_ARCHITECTURE_ID`, which CMake sets automatically based on detecting things from the compiler. However this is set only in some cases so one can't generally depend on it either, see https://gitlab.kitware.com/cmake/cmake/-/issues/17702 and https://gitlab.kitware.com/cmake/cmake/-/merge_requests/6432.

So TL;DR if you want something that automatically detects this from the compiler rather than having the user specify it, then something like the suggested example probably is quite reasonable.

Reply:
I forgot to add the patch for Tests/CMakeLists.txt where the ARCHITECTURE variable is actually used.

Reply:
This, indeed, sounds like a problem with your cross-compile setup.  If CMake is setting the wrong architecture type then all manner of other things will be broken.

Reply:
I am not cross-compiling in the traditional sense. I can either use a AMD64 or ARM64 version of CMake on WoA. Most distributions only provide a AMD64 release. The Build System is not a native process, which is fine.

Despite the difference, CMake selects the native version of LLVM (AArch64).

As @mstorsjo said:
> So TL;DR if you want something that automatically detects this from the compiler rather than having the user specify it, then something like the suggested example probably is quite reasonable.

It is hard to detect a non-native CMake without my patch. Force users to use the ARM64 version, and tolerate weird behaviour when running a non-native version?

What is your solution?

Reply:
Can you report this upstream to CMake?  If it's finding the AArch64 tools, it should know that the target is AArch64.  I presume that there's a similar issue using an x86-64 binary under Rosetta 2 on macOS when targeting the native platform.

Reply:
https://gitlab.kitware.com/cmake/cmake/-/issues/25466